src/H5FDint.c

```diff
@@ -1248,3 +1333,258 @@ H5FD_driver_query(const H5FD_class_t *driver, unsigned long *flags /*out*/)
+
+    /* Sanity checks */
+    HDassert(vector_was_sorted);
+
+    HDassert((types) || (count == 0));
+    HDassert((addrs) || (count == 0));
+    HDassert((sizes) || (count == 0));
+    HDassert((bufs) || (count == 0));
+
+    /* verify that the first elements of the sizes and types arrays are
+     * valid.
+     */
+    HDassert((count == 0) || (sizes[0] != 0));
+    HDassert((count == 0) || (types[0] != H5FD_MEM_NOLIST));
+
+    HDassert((count == 0) || ((s_types_ptr) && (NULL == *s_types_ptr)));
+    HDassert((count == 0) || ((s_addrs_ptr) && (NULL == *s_addrs_ptr)));
+    HDassert((count == 0) || ((s_sizes_ptr) && (NULL == *s_sizes_ptr)));
+    HDassert((count == 0) || ((s_bufs_ptr) && (NULL == *s_bufs_ptr)));
+
+    *vector_was_sorted = TRUE;
+
+    /* if count <= 1, vector is sorted by definition */
+    if (count > 1) {
+
+        /* scan the addrs array to see if it is sorted */
+        i = 1;
+
+        while ((*vector_was_sorted) && (i < (int)(count - 1))) {
+
+            if (H5F_addr_gt(addrs[i - 1], addrs[i])) {
+
+                *vector_was_sorted = FALSE;
+            }
+            else if (H5F_addr_eq(addrs[i - 1], addrs[i])) {
+
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "duplicate addr in vector")
+            }
+            i++;
+        }
+    }
+
+    if (*vector_was_sorted) {
+
+        *s_types_ptr = types;
+        *s_addrs_ptr = addrs;
+        *s_sizes_ptr = sizes;
+        *s_bufs_ptr  = bufs;
+    }
+    else {
+
+        /* must sort the addrs array in increasing addr order, while
+         * maintaining the association between each addr, and the
+         * sizes[], types[], and bufs[] values at the same index.
+         *
+         * Do this by allocating an array of struct H5FD_vsrt_tmp_t, where
+         * each instance of H5FD_vsrt_tmp_t has two fields, addr and index.
+         * Load the array with the contents of the addrs array and
+         * the index of the associated entry. Sort the array, allocate
+         * the s_types_ptr, s_addrs_ptr, s_sizes_ptr, and s_bufs_ptr
+         * arrays and populate them using the mapping provided by
+         * the sorted array of H5FD_vsrt_tmp_t.
+         */
+        int    j;
+        int    fixed_size_index = (int)count;
+        int    fixed_type_index = (int)count;
+        size_t srt_tmp_size;
+
+        srt_tmp_size = ((size_t)count * sizeof(struct H5FD_vsrt_tmp_t));
+
+        if (NULL == (srt_tmp = (H5FD_vsrt_tmp_t *)HDmalloc(srt_tmp_size)))
+
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "can't alloc srt_tmp")
+
+        for (i = 0; i < (int)count; i++) {
+
+            srt_tmp[i].addr  = addrs[i];
+            srt_tmp[i].index = i;
+        }
+
+        /* sort the srt_tmp array */
+        HDqsort(srt_tmp, (size_t)count, sizeof(struct H5FD_vsrt_tmp_t), H5FD__vsrt_tmp_cmp);
+
+        /* verify no duplicate entries */
+        i = 1;
+
+        while ((*vector_was_sorted) && (i < (int)(count - 1))) {
```

Comment:
Won't *vector_was_sorted always be false here?

Reply:
You are correct --  mindless cut and paste.  Will fix.

Reply:
I also wonder if we need to be doing the work of scanning the array once again while looking for duplicates. I could be wrong, but I think MPI I/O doesn't care much about duplicates and you'd only get in trouble if you were trying to write different data to the same offset.

Of course, it may not quite be correct to just omit checking for duplicates, but we also might be erroring on somebody's I/O call that may not be problematic. Thoughts on that?

Reply:
My impression is that MPI does not like overlapping entries in the derived types passed to MPI_File_set_view -- at least I triggered a number of failures by accident that seemed to indicate this. 

Even if MPI is OK with overlapping entries, wouldn't it be an error if we write the same location twice in a single I/O request?  If not, your argument against this check may be sound -- but I can't think of an example.

Reply:
This very well may be the case, though MPI's "monotonically non-decreasing" requirement seems to hint that this may be a possibility. I vaguely recall running into this with parallel compression at one point, but that was long enough ago that I don't remember the details. I would think that yes, writing the same location twice wouldn't generally be correct, but as long as the data is the exact same I wouldn't think it would be problematic.

In any case, I was mostly looking at the overall read/write process to see if it's possible to cut down on any work. It seems that, on average, for each I/O request we're likely going to:

- scan through the whole addrs array to check if it's sorted
- allocate and populate the H5FD_vsrt_tmp_t array
- sort the H5FD_vsrt_tmp_t array and scan it for duplicates
- allocate copies of the types, addrs, sizes, bufs arrays
- scan the sizes and types vectors for optimizations

This is probably just a case of premature optimization, but it seemed like a lot of work was going on, so I just wondered if there are places we can cut corners or not.